### DISCUSSION TITLE
How check own routes before checking leptos ones?

### DISCUSSION THREAD

**Discussion by Umenokin** [leptos-rs/leptos#1913] [Q&A] [answered]
I am trying to provide CMS-like routing like:

1. `/* - user pages`
2. `/admin/* - admin dashboard`

## My current solution

While ideally, I would prefer to check user pages first I managed to make a working example based on Leptos's examples where:
1. Check admin (Leptos) routes
2. If not Leptos routes, got to fallback
3. Return admin file asset if one was found
4. Check the user page (`runtime_handler` is the one serving those)
5. Return error

```
// Routes
    let app = Router::new()
        .leptos_routes(&leptos_options, routes, App)
        .fallback(runtime_and_other_handlers)
        .with_state(leptos_options);
        
// Fallback
pub async fn runtime_and_other_handlers(
    uri: Uri,
    State(leptos_options): State<LeptosOptions>,
    req: Request<Body>,
) -> AxumResponse {
    let root = leptos_options.site_root.clone();
    let res = get_static_file(uri.clone(), &root).await.unwrap();

    if res.status() == StatusCode::OK {
        res.into_response()
    } else {
        let res = runtime_handler(&req).await;
        if res.status() == StatusCode::OK {
            return res;
        }

        let mut errors = Errors::default();
        errors.insert_with_default_key(AppError::NotFound);
        let handler = leptos_axum::render_app_to_stream(
            leptos_options.to_owned(),
            move || view! { <ErrorTemplate outside_errors=errors.clone()/> },
        );
        handler(req).await.into_response()
    }
}
```

It would be great if I could do it in the following order:
1. Check the user page (`runtime_handler` is the one serving those)
2. If not check admin (Leptos) routes 
3. Go to fallback

The Axum group could be a better place to ask this question, but in relation to Leptos, how can I pass the custom state to the fallback if I continue with my working solution? Right now `with_state` allow passing only `LeptosOptions`

**Comment by Umenokin** [comment, marked as answer]
Thanks to the Axum team's I managed to make a solution that seems works well so far:
```
    let app = Router::new()
        .route("/*path", any(|| async { Html("Hello, World!") }))
        .route("/pkg/*path", get(handle_static_files))
        .route("/admin/assets/*path", get(handle_static_files))
        .leptos_routes(&leptos_options, routes, App)
        .with_state(leptos_options);
```

The only thing I would've wanted to be able to do is to move `pkg` path under the `/admin` but was not able to find the right configuration. With the configuration below I am able to build wasm, js, and CSS in the right folder and I can serve CSS from it, wasm and js files still will have uri `/pkg/bundle.js` instead of `/admin/pkg/bundle.js`.

```
site-root = "target/site/admin"
site-pkg-dir = "pkg"
```

Also, I was able to pass custom state like so:
```
async fn leptos_routes_handler(State(app_state): State<AppState>, req: Request<AxumBody>) -> Response {
    let handler = leptos_axum::render_route(
        app_state.leptos_options.clone(),
        app_state.routes.clone(),
        App
    );

    handler(req).await.into_response()
}

...

    let app = Router::new()
        .route("/*path", any(|| async { Html("Hello, World!") }))
        .route("/pkg/*path", get(handle_static_files))
        .route("/admin/assets/*path", get(handle_static_files))
        .leptos_routes_with_handler(routes, get(leptos_routes_handler) )
        .with_state(app_state);
```

**Reply by gbj** [reply]
That's great! My advice would have been something very much along the same lines, so I'm glad you were able to figure it out.

**Reply by Umenokin** [reply]
Confirming the right way is also very important for me. Thank you, @gbj!
